### DISCUSSION TITLE
Is it possible to end a span created from context propagation?

### DISCUSSION THREAD

**Discussion by MrRickwi** [open-telemetry/opentelemetry-java#3388] [Q&A]
We are instrumenting a workflow engine with trace support. We have a plugin that listens for the START and END events of a workflow. In the START event, we open a span and inject this into the workflow. In the END event, we would like to extract the injected span (traceparent, ...) from the workflow and end this one.

However, the following code seems not work maybe due to the fact that the span we get there is an ImmutableSpan.

START:
```
      Span span = tracer.spanBuilder(execution.getProcessDefinitionId() + ":" + execution.getProcessInstanceId()).startSpan();
      // Insert the new span into the current workflow
      try (Scope scope = span.makeCurrent()) {
        openTelemetry.getPropagators().getTextMapPropagator().inject(Context.current(), execution, TextMapInjectAdapter.SETTER);
      }
```

END:
```
    Context extractedContext = openTelemetry.getPropagators().getTextMapPropagator().extract(Context.current(), execution, TextMapExtractAdapter.GETTER);
    try (Scope extractedScope = extractedContext.makeCurrent()) {
      Span span = Span.fromContext(extractedContext);
      span.end();
    }
```

Any ideas around there? Remembering the span in code is not feasible because the workflow engine may be restarted while the workflow is running, thus loosing any transinet information in the workflow.

**Comment by jkwatson** [comment]
Hi @MrRickwi . You are correct in finding that this is not possible. In OpenTelemetry, propagated spans are assumed to have their lifecycle fully managed by the creating process. In your case, you probably need to model this as two separate spans, connected either via standard parenting, or via a span Link, depending on how you would like to see the structure.

**Reply by MrRickwi** [reply]
Thanks @jkwatson for the reply: So, my problem at least is not to influence propagated spans, but to open a span in some context (START workflow), save it's properties somewhere (the instance of the executed workflow), pass that span as propagated context to some code executed inside the workflow (which works exactly with inject/extract), and close that span based on that saved properties (in the instance of the same workflow) in some other context (END workflow) typically a long time later.

**Reply by MrRickwi** [reply]
Here is what I actually do and what makes the workflow instance not stateless:

```
public class TracingExecutionListener implements ExecutionListener {

  private final Logger LOGGER = Logger.getLogger(this.getClass().getName());
  private final OpenTelemetry openTelemetry;
  private final HashMap<String, Span> activeSpans = new HashMap<String, Span>();
  private static final String INSTRUMENTATION_NAME = TracingExecutionListener.class.getName();

  public TracingExecutionListener(OpenTelemetry openTelemetry) {
    this.openTelemetry = openTelemetry;
  }

  @Override
  public void notify(DelegateExecution execution) throws Exception {
    LOGGER.info(this.getClass().getName() + ".notify(" + execution.getEventName() + "): " + execution.getProcessDefinitionId() + ":" + execution.getProcessInstanceId());
    Tracer tracer = openTelemetry.getTracer(INSTRUMENTATION_NAME);
    switch (execution.getEventName()) {
      case ExecutionListener.EVENTNAME_START:
        start(tracer, execution);
        break;
      case ExecutionListener.EVENTNAME_END:
        end(tracer, execution);
        break;
    }
  }

  private void start(Tracer tracer, DelegateExecution execution) {
    // Extract the SpanContext and other elements from the workflow
    Context extractedContext = openTelemetry.getPropagators().getTextMapPropagator().extract(Context.current(), execution, TextMapExtractAdapter.GETTER);
    try (Scope extractedScope = extractedContext.makeCurrent()) {
      // Create span and remember as open span
      Span span = tracer.spanBuilder(execution.getProcessDefinitionId() + ":" + execution.getProcessInstanceId()).startSpan();
      activeSpans.put(execution.getProcessInstanceId(), span);
      // Inject the new span into the current workflow
      try (Scope scope = span.makeCurrent()) {
        openTelemetry.getPropagators().getTextMapPropagator().inject(Context.current(), execution, TextMapInjectAdapter.SETTER);
      }
    }
  }

  private void end(Tracer tracer, DelegateExecution execution) {
    // Get the active span from memory (not stateless!)...
    if (activeSpans.containsKey(execution.getProcessInstanceId())) {
      Span span = activeSpans.get(execution.getProcessInstanceId());
      activeSpans.remove(execution.getProcessInstanceId());
      // ...and end that span
      span.end();
    }
  }

}
```

**Reply by MrRickwi** [reply]
Possibly the best solution would be to save the serialized span in the workflow and to deserialize the span at the end and end it there.

**Reply by jkwatson** [reply]
Is there a reason you can't just have the start and end be two different spans, with start being the parent of end?

**Reply by MrRickwi** [reply]
@jkwatson, in my first experiments I created a span in START, injected it into my workflow context, and ended it there. As a result, I saw only the start span and no child spans, I assume due to the fact that the injected start span already was ended when new spans where added as child. Then I tried never to close the start span, and as a result I see nothing - makes sense, the span is still active and not shown in Jaeger. So that is the reason why I try to end the start span exactly where it should end - at the end of the workflow. If there is another approach, I would appreciate it.

**Reply by jkwatson** [reply]
As I said..I think your only approach here is to have two spans. One for the start of the process and one for the end of the process. Inject the start-of-process span into an execution context of some sort, then extract it and use it as the parent of the span that tracks when your process gets ended. OpenTelemetry doesn't support your use-case as you want to do it. If the end of the process might take place on a different machine/JVM, then you have to have it be a 2nd span.

**Reply by MrRickwi** [reply]
So I will give it a try. Last question: In your approach, would you end the first span after injecting it into the execution context?

**Reply by jkwatson** [reply]
I think so, yes. Effectively the span is "start the job", and once it's been kicked off and submitted, then the span would end.
